### PR TITLE
Support changing domain_join settings

### PR DESCRIPTION
**Pull request by mikumikuchan**
This PR is based on https://github.com/vijos/vj4/pull/353.

### CODE REVIEW

File: vj4/handler/domain.py [hunk 1]
```diff
@@ -93,6 +97,52 @@ class DomainEditHandler(base.Handler):
     self.json_or_redirect(self.url)
 
 
+@app.route('/domain/join_applications', 'domain_manage_join_applications')
+class DomainJoinApplicationsHandler(base.Handler):
+  @base.require_perm(builtin.PERM_EDIT_DESCRIPTION)
```
Comment (twd2) [inline vj4/handler/domain.py]: `PERM_EDIT_PERM`?

File: vj4/handler/domain.py [hunk 2]
```diff
@@ -93,6 +97,52 @@ class DomainEditHandler(base.Handler):
     self.json_or_redirect(self.url)
 
 
+@app.route('/domain/join_applications', 'domain_manage_join_applications')
```
Comment (twd2) [inline vj4/handler/domain.py]: `join_settings`?

File: vj4/handler/domain.py [hunk 3]
```diff
@@ -93,6 +97,52 @@ class DomainEditHandler(base.Handler):
...
+class DomainJoinApplicationsHandler(base.Handler):
+  @base.require_perm(builtin.PERM_EDIT_DESCRIPTION)
+  async def get(self):
+    roles = sorted(list(self.domain['roles'].keys()))
+    roles_with_text = [(role, role) for role in roles]
```
Comment (twd2) [inline vj4/handler/domain.py]: ???

File: vj4/handler/domain.py [hunk 4]
```diff
@@ -93,6 +97,52 @@ class DomainEditHandler(base.Handler):
...
+    join_settings = domain.get_join_settings(self.domain)
+    expirations = [(k, v) for k, v in vj4.constant.domain.JOIN_EXPIRATION_RANGE.items()]
```
Comment (twd2) [inline vj4/handler/domain.py]: `list(vj4.constant.domain.JOIN_EXPIRATION_RANGE.items())`?

File: vj4/handler/domain.py [hunk 5]
```diff
@@ -93,6 +97,52 @@ class DomainEditHandler(base.Handler):
...
+    if join_settings is None:
```
Comment (twd2) [inline vj4/handler/domain.py]: `if not join_settings`

File: vj4/handler/domain.py [hunk 6]
```diff
@@ -93,6 +97,52 @@ class DomainEditHandler(base.Handler):
...
+      expirations = expirations[1:]
```
Comment (twd2) [inline vj4/handler/domain.py]: ???

File: vj4/handler/domain.py [hunk 7]
```diff
@@ -93,6 +97,52 @@ class DomainEditHandler(base.Handler):
...
+    self.render('domain_manage_join_applications.html', roles_with_text=roles_with_text,
+                join_settings=join_settings, expirations=expirations)
+
+  @base.require_perm(builtin.PERM_EDIT_DESCRIPTION)
```
Comment (twd2) [inline vj4/handler/domain.py]: ditto?

File: vj4/handler/domain.py [hunk 8]
```diff
@@ -93,6 +97,52 @@ class DomainEditHandler(base.Handler):
...
+  @base.post_argument
+  @base.require_csrf_token
+  @base.sanitize
+  async def post(self, *, method: int, role_assignment: str=None, expire: int=None,
```
Comment (twd2) [inline vj4/handler/domain.py]: Just call it `role`?

File: vj4/handler/domain.py [hunk 9]
```diff
@@ -93,6 +97,52 @@ class DomainEditHandler(base.Handler):
...
+    if method not in constant.domain.JOIN_METHOD_RANGE:
+      raise error.ValidationError('method')
+    if method == constant.domain.JOIN_METHOD_NONE:
+      join_settings = None
+    else:
+      if role_assignment not in self.domain['roles']:
+        raise error.ValidationError('role_assignment')
+      if expire not in constant.domain.JOIN_EXPIRATION_RANGE:
+        raise error.ValidationError('expire')
+      if current_join_settings is None and expire == constant.domain.JOIN_EXPIRATION_KEEP_CURRENT:
```
Comment (twd2) [inline vj4/handler/domain.py]: `if not current_join_settings and`...

File: vj4/handler/domain.py [hunk 10]
```diff
@@ -93,6 +97,52 @@ class DomainEditHandler(base.Handler):
...
+      if method == constant.domain.JOIN_METHOD_CODE:
+        join_settings['code'] = invitation_code
+      if expire == constant.domain.JOIN_EXPIRATION_KEEP_CURRENT:
+        join_settings['expire'] = current_join_settings['expire']
+      elif expire == constant.domain.JOIN_EXPIRATION_UNLIMITED:
+        join_settings['expire'] = None
+      else:
+        join_settings['expire'] = datetime.datetime.utcnow() + datetime.timedelta(hours=expire)
+    await domain.edit(self.domain_id, join=join_settings)
+    self.json_or_redirect(self.url)
```
Comment (twd2) [inline vj4/handler/domain.py]: `referer_or_main`?

File: vj4/model/domain.py [hunk 11]
```diff
@@ -214,6 +217,21 @@ def get_multi_user(*, fields=None, **kwargs):
   return result
 
 
+def get_join_settings(ddoc):
+  if 'join' not in ddoc:
+    return None
+  join_settings = ddoc['join']
+  if join_settings is None:
```
Comment (twd2) [inline vj4/model/domain.py]: `if not `...

File: vj4/model/domain.py [hunk 12]
```diff
@@ -214,6 +217,21 @@ def get_multi_user(*, fields=None, **kwargs):
...
+    return None
+  if join_settings['method'] == constant.domain.JOIN_METHOD_NONE:
+    return None
+  if join_settings['role'] not in ddoc['roles']:
+    return None
+  if join_settings['expire'] != None and join_settings['expire'] < datetime.datetime.utcnow():
```
Comment (twd2) [inline vj4/model/domain.py]: `if join_settings['expire'] and join_settings['expire'] <`...

File: vj4/util/validator.py [hunk 13]
```diff
@@ -11,6 +11,7 @@
 DOMAIN_ID_RE = re.compile(r'[_A-Za-z][_0-9A-Za-z]{3,255}')
 ID_RE = re.compile(r'[^\\/\s\u3000]([^\\/\n\r]*[^\\/\s\u3000])?')
 ROLE_RE = re.compile(r'[_0-9A-Za-z]{1,256}')
+DOMAIN_INVITATION_CODE_RE = re.compile(r'[0-9A-Za-z]{1,20}')
```
Comment (twd2) [inline vj4/util/validator.py]: `INVITATION_CODE_RE`

File: vj4/util/validator.py [hunk 14]
```diff
@@ -142,3 +143,12 @@ def is_lang(i):
 def check_lang(i):
   if not is_lang(i):
     raise error.ValidationError('lang')
+
+
+def is_domain_invitation_code(s):
```
Comment (twd2) [inline vj4/util/validator.py]: `is_invitation_code`

File: vj4/util/validator.py [hunk 15]
```diff
@@ -142,3 +143,12 @@ def is_lang(i):
...
+  return bool(DOMAIN_INVITATION_CODE_RE.fullmatch(s))
+
+
+def check_domain_invitation_code(s):
```
Comment (twd2) [inline vj4/util/validator.py]: `check_invitation_code`

and unit test cases, please.

File: vj4/model/domain.py (hunk 12 above)
Comment (breeswish) [reply vj4/model/domain.py]: they are not equal?

File: vj4/util/validator.py (hunk 15 above)
Comment (breeswish) [reply vj4/util/validator.py]: The regular expression that function used is so simple that it's correctness doesn't need to be verified.

File: vj4/handler/domain.py (hunk 8 above)
Comment (breeswish) [reply vj4/handler/domain.py]: more accurate naming is not bad.

File: vj4/handler/domain.py (hunk 1 above)
Comment (breeswish) [reply vj4/handler/domain.py]: Done

File: vj4/model/domain.py [hunk 16]
```diff
@@ -214,6 +217,21 @@ def get_multi_user(*, fields=None, **kwargs):
   return result
 
 
+def get_join_settings(ddoc):
```
Comment (twd2) [inline vj4/model/domain.py]: Pass `now` as an argument.

**Pull request merged**